components/omega/src/base/MachEnv.cpp

```diff
@@ -0,0 +1,441 @@
+
+   // Define the range of tasks in new group (0, NewSize-1)
+   int NRanges     = 1;
+   int LastTask    = NewSize - 1;
+   int Range[1][3] = {0, LastTask, 1};
+
+   // Create a new group with the new range
+   MPI_Group NewGroup;
+   MPI_Group_range_incl(InGroup, NRanges, Range, &NewGroup);
+
+   // Create the communicator for the new group
+   MPI_Comm_create(InComm, NewGroup, &NewEnv.Comm);
+
+   // determine whether this task is a part of the new communicator
+   if (NewEnv.Comm != MPI_COMM_NULL) {
+
+      NewEnv.MemberFlag = true;
+      // get task ID for local MPI task
+      MPI_Comm_rank(NewEnv.Comm, &NewEnv.MyTask);
+      // get total number of MPI tasks
+      MPI_Comm_size(NewEnv.Comm, &NewEnv.NumTasks);
+
+   } else {
+
+      NewEnv.MemberFlag = false;
+      NewEnv.MyTask     = -999;
+      NewEnv.NumTasks   = -999;
+   }
+
+   // Set task 0 as master
+   NewEnv.MasterTask = 0;
+
+   // determine if this task is the master task
+   if (NewEnv.MyTask == NewEnv.MasterTask) {
+      NewEnv.MasterTaskFlag = true;
+   } else {
+      NewEnv.MasterTaskFlag = false;
+   }
+
+#ifdef OMEGA_THREADED
+   // total number of OpenMP threads
+   NewEnv.NumThreads = omp_get_num_threads();
+#else
+   NewEnv.NumThreads = 1;
+#endif
+
+   // Add to list of environments
+   OMEGA::MachEnv::addEnv(Name, &NewEnv);
+
+} // end create with contiguous range
+
+//------------------------------------------------------------------------------
+// Create a new environment from a strided subset of an
+// existing environment
+
+void MachEnv::createEnv(const std::string Name, // [in] name of environment
+                        const MachEnv *InEnv,   // [in] parent MachEnv
+                        const int NewSize,      // [in] num tasks in new env
+                        const int Begin,        // [in] starting parent task
+                        const int Stride        // [in] stride for tasks to incl
+) {
+
+   // Create the new Env structure and make it persistent (static)
+   static MachEnv NewEnv;
+
+   // First retrieve the group associated with the parent environment
+   MPI_Comm InComm = InEnv->getComm();
+   MPI_Group InGroup;
+   MPI_Comm_group(InComm, &InGroup);
+
+   // Define the range of tasks in new group based on the input
+   // start task and stride
+   int NRanges     = 1;
+   int LastTask    = Begin + (NewSize - 1) * Stride;
+   int Range[1][3] = {Begin, LastTask, Stride};
+
+   // Error checks for valid range
+   int OldSize;
+   MPI_Comm_size(InComm, &OldSize);
+   if (LastTask >= OldSize) {
+      std::cerr << "Invalid range in strided MachEnv constructor "
+                << "LastTask = " << LastTask << " is larger than old size"
+                << OldSize << std::endl;
+      return;
+   }
+
+   // Create a new group with the new range
+   MPI_Group NewGroup;
+   MPI_Group_range_incl(InGroup, NRanges, Range, &NewGroup);
+
+   // Create the communicator for the new group
+   MPI_Comm_create(InComm, NewGroup, &(NewEnv.Comm));
```

Comment:
Perhaps we can do a error check for comm_create so that it exits here than somewhere further down stream.

Reply:
I can add, but it turns out that there are few instances where an actual error is returned. Most often, it will return a MPI_COMM_NULL which can be for a variety of valid reasons.